decompiler/frontend/binaryninja/handlers/constants.py

```diff
@@ -77,26 +82,65 @@ def _lift_symbol_pointer(self, address: int, symbol: bSymbol) -> Optional[Symbol
+                OperationType.address,
+                [GlobalVariable(variable_name, ssa_label=0, vartype=vartype, initial_value=addr)],
+                vartype=Pointer(vartype),
+            )
+        if parent_addr == addr:
+            # We have cases like:
+            # void* __dso_handle = __dso_handle
+            # Prevent unlimited recursion and return the pointer.
+            vartype = Integer.uint64_t() if bv.address_size == 8 else Integer.uint32_t()
+            return GlobalVariable(variable_name, vartype=vartype, ssa_label=0, initial_value=addr)
+
+        # Retrieve the initial value of the global variable if there is any
+        type_tokens = [t.text for t in variable.type.tokens]
+        if variable.type == variable.type.void():
+            # If there is no type, just retrieve all the bytes from the current to the next address where a data variable is present.
+            next_data_var_addr = bv.get_next_data_var_after(addr).address
+            num_bytes = next_data_var_addr - addr
+            initial_value = bv.read(addr, num_bytes)
+            initial_value = self._get_bytes(bv, addr)
+        elif variable.type.type_class == TypeClass.IntegerTypeClass:
+            initial_value = self._get_integer(bv, addr, variable.type.width)
+        else:
+            # Handle general case
+            type_width = variable.type.width
+            tmp_value = bv.read(addr, type_width)
+            # If pointer type, convert tmp_value to a label, otherwise leave it as it is.
+            if "*" in variable.type.tokens:
+                indirect_ptr_addr = int.from_bytes(tmp_value, ConstantHandler.Endian[bv.endianness])
+                initial_value = self._lift_global_variable(bv, indirect_ptr_addr)
+            # If pointer type, convert indirect_pointer to a label, otherwise leave it as it is.
+            if "*" in type_tokens:
+                indirect_ptr_addr = self._get_pointer(bv, addr)
+                initial_value = self._lift_global_variable(bv, addr, indirect_ptr_addr)
+            else:
+                initial_value = tmp_value
+                initial_value = bv.read(addr, variable.type.width)
+        # Create the global variable.
+        # Convert all void and void* to char*
+        if "void" in type_tokens:
+            vartype = self._lifter.lift(bv.parse_type_string("char*")[0])
         return UnaryOperation(
             OperationType.address,
-            [
-                GlobalVariable(
-                    variable.name if variable.name else f"data_{hex(symbol.address)}",
-                    vartype=self._lifter.lift(variable.type),
-                    ssa_label=0,
-                    initial_value=initial_value,
-                )
-            ],
-            vartype=Pointer(self._lifter.lift(variable.type)),
+            [GlobalVariable(variable_name, vartype=vartype, ssa_label=0, initial_value=initial_value)],
+            vartype=Pointer(vartype),
         )
```

Comment:
This function is very long and does a lot. Could you refactor it into multiple functions (and maybe classes)? I am a bit afraid that when Binja again changes some stuff, it would be hard to change this. 
Perhaps write small functions explaining the conditions of if-statements, try to have at least one function per if-block to explain what we do. Afterward, we can see how long the function is and whether it should be split into even more functions. 

One other question, are the TODOs covered in another issue? Or are they no TODOs and are they only something we assume?

Reply:
I have tried to refactor it. Hopefully it looks better now! 

As for the TODOs, I believe they appear to be upstream issues within binaryninja that we can't fix. In that case, I'm not sure if we want to have issues for those since those aren't within our control.